### PR TITLE
Strict pyrefly checking: round 1 (4/N)

### PR DESCRIPTION
**Pull request by aorenste**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #191253

Annotates and turns on whole-file pyrefly checking (per-file [[sub-config]]
entries) for torch/optim/{optimizer,adamw,_functional}.py. Fourth batch of the
effort to enable annotation checking on non-public torch files.

These are the only three optim files that qualify: the concrete optimizers
(adam, sgd, ...) and Adafactor/Muon expose a public step() re-exported via
torch.optim.__all__, so turning on strict checking for them is out of scope.

Most of the diff gives optimizer.py's shared numeric helpers real types instead
of Any: _get_value and _to_scalar are overloaded (Tensor->Tensor / float->float),
_stack_if_compiling returns torch.Tensor | Sequence[torch.Tensor | float], and
_cast is a structure-preserving _T -> _T TypeVar.

Typing them surfaced a latent annotation bug: several _single_tensor_*/
_multi_tensor_* helpers declared lr: float even though tensor LR is supported
(the code branches on isinstance(lr, torch.Tensor)) -- the Any returns had
masked it. Those params are widened to float | Tensor to match the convention
adam.py already uses. Where a possibly-tensor lr then reaches a scalar-only
Tensor API (add_ alpha=, addcdiv_ value=, _foreach_*), a scoped
`# type: ignore[arg-type]` documents the eager/compile duality (scalar in eager,
tensor under torch.compile) that no static type expresses -- again the pattern
adam.py already uses.

Test Plan:

```
lintrunner -a torch/optim/*.py
pyrefly check   # whole project: 156 errors == baseline, no net-new
```

The changes are type-only (annotations, casts, # type: ignore comments, and a
semantically-identical rewrite of adam's betas tuple construction) with no
runtime behavior change. test/test_optim.py was not run here because the local
torch._C extension is stale (unrelated missing symbols); run it after a rebuild.

Authored with Claude.

### CODE REVIEW

**Review by Skylion007 — APPROVED** [review]

File: torch/optim/optimizer.py [hunk 1]
```diff
@@ -87,17 +86,17 @@ def _use_grad(*args: _P.args, **kwargs: _P.kwargs) -> _T:
     return _use_grad
 
 
-def _get_value(x):
+def _get_value(x: torch.Tensor | float) -> Any:
```
Comment (Skylion007) [inline torch/optim/optimizer.py]: meh this could be slightly more specific with a bound typevar?

**Comment by Skylion007** [comment]
@claude review this PR, focusing on ways typing could be further improved.
